[BEAM-14337] Support **kwargs for PyTorch models

yeandy opened the pull request:
We currently only support ability to pass in a single Tensor that does not have any keys attached to it: `model(input)`. We want to support kwargs like this: `model(**inputs)`.

Because of the way Beam returns `BatchElements()`, we need to convert a list of (key to Tensor) dicts into a dict of Tensor lists.

This PR also adds support for non-batchable params that are passed in as side inputs.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

yeandy commented:
R: @AnandInguva

AnandInguva reviewed (approved):
1 comment. LGTM

TheNeuralBit commented on sdks/python/apache_beam/ml/inference/pytorch.py:
```diff
@@ -45,11 +46,22 @@ def run_inference(self, batch: List[torch.Tensor],
-    if batch.device != self._device:
-      batch = batch.to(self._device)
-    predictions = model(batch)
+    if isinstance(batch[0], dict):
```
It might clean things up to only handle the general case and not have special logic for the single-argument case.

robertwb replied:
+1. If PytorchInferenceRunner (and its associated ModelLoder) were typed, we would have one that takes tensors and one that takes dicts of tensors, and this typing would go all the way through to the PCollections being PCollections of tensors vs. PCollections of dicts of tensors.

yeandy commented:
PTAL @AnandInguva @ryanthompson591

ryanthompson591 commented on sdks/python/apache_beam/ml/inference/base.py:
```diff
@@ -96,7 +107,9 @@ def expand(self, pcoll: beam.PCollection) -> beam.PCollection:
-        | beam.ParDo(_RunInferenceDoFn(self._model_loader, self._clock)))
+        | beam.ParDo(
+            _RunInferenceDoFn(
+                self._model_loader, self._prediction_params, self._clock)))
```
This is also what I preferred. But after a long discussion we settled on making these a side input.

The reason was that since these static parameters were conceptually different because the were parameters for the inputs rather than model loading parameters, they should not be passed in on the model loader. If you want to rehash this I suggest setting up a meeting.

yeandy replied:
Note: The code has been changed slightly to `**kwargs` instead of `self._prediction_params`.

Passing in these extra prediction-related parameters as side inputs makes more conceptual sense because both the `batch` examples and `prediction_params` kwargs are different types of inputs that are both passed at the same time into Pytorch model's predict call (which are both passed into its `forward` function)
```
# Inside PytorchInferenceRunner()
def run_inference(batch, model, **kwargs):
   prediction_params = kwargs.get('prediction_params')
   model(batch, **prediction_params)

# this call goes inside Pytorch model ...
class PytorchModel(nn.Module):
    def forward(self, batch, **prediction_params):
        ...
```

TheNeuralBit replied:
> But after a long discussion we settled on making these a side input.

This is not a side input, this is a constant parameter. A side input comes from a PCollection wrapped in a PCollectionView and is passed to the ParDo: https://beam.apache.org/documentation/programming-guide/#side-inputs

That's what I'm trying to suggest, making it a side input would justify a change to base RunInference, because it means you'd need to change the `ParDo` creation.

yeandy replied:
Thanks for the clarification Brian. 

It may be the case that these extra `prediction_params` need to be determined at runtime (not hard-coded) based on, for example, the shape of the input data. So I think we'd want to modify base.. I'm going to do a few quick tests right now to see the flow of passing in `prediction_params` as a PCollection.

ryanthompson591 commented on sdks/python/apache_beam/ml/inference/pytorch.py:
```diff
@@ -39,25 +42,47 @@ class PytorchInferenceRunner(InferenceRunner):
     the inference call.
     """
+    if prediction_params is None:
+      prediction_params = {}
```
I don't see the need for this. If prediction params is none this test should still work right?

Prediction params should be optional right?

yeandy commented on sdks/python/apache_beam/ml/inference/base_test.py:
```diff
@@ -39,7 +41,12 @@ class FakeInferenceRunner(base.InferenceRunner):
+      self,
+      batch: Any,
+      model: Any,
+      prediction_params: Optional[Dict[str, Any]] = None,
```
If I take out `prediction_params`, I get 
```
      result_generator = self._inference_runner.run_inference(
>         examples, self._model, self._prediction_params)
E     TypeError: run_inference() takes 3 positional arguments but 4 were given [while running 'RunInference/ParDo(_RunInferenceDoFn)']

apache_beam/ml/inference/base.py:216: TypeError
```

If we allow this not to be a required param, then maybe we can check if `_prediction_params` is `None` like so? 
```
    if not self._prediction_params:
        result_generator = self._inference_runner.run_inference(
            examples, self._model, self._prediction_params)
    else:
            result_generator = self._inference_runner.run_inference(
            examples, self._model)
```
That feels a bit messy though, and breaks our `run_inference` interface, assuming we want to use this modification:
```
class InferenceRunner():
  def run_inference(
      self,
      batch: List[Any],
      model: Any,
      prediction_params: Optional[Dict[str, Any]] = None) -> Iterable[Any]:
    """Runs inferences on a batch of examples and
    returns an Iterable of Predictions."""
    raise NotImplementedError(type(self))
```

ryanthompson591 replied:
Heres some code to demonstrate what I mean.

def pytorch_impl(prediction_params = None):
  print ('got prediction_params ' + str(prediction_params))

def tf_impl():
  # Does not have the prediction_params arg, should not crash when called
  # but should crash if called with prediction_params argument
  print('tf_impl')

mode = 'tf'

def base_do_fn(**kwargs):
  if mode == 'tf':
    tf_impl(**kwargs)
  elif mode == 'pytorch':
    pytorch_impl(**kwargs)


def base_run_inference(**kwargs):
  base_do_fn(**kwargs)

#passes
mode='tf'
base_run_inference()

#passes
mode='pytorch'
base_run_inference(prediction_params={'prediction_param': 'value'})

# fails (expected), because we're passing a parameter that TF implementation should not get
mode='tf'
base_run_inference(prediction_params={'prediction_param': 'value'})

----------------
What I'm saying is this method allows forwarding of any named param whatsoever. With what you're doing we'll always be in a spot of modifying an interface and thus having to modify every implementation of the interface going forward for any single parameter that any single implementation wants.

I think here, and going forward, unless named parameters will apply to all models we should not put it into the interface for all models.

yeandy replied:
Ah I see, thanks for the clarification. I was thinking of something else, hence the misunderstanding. I've added `kwargs` to support generalizability.

yeandy commented on sdks/python/apache_beam/ml/inference/pytorch.py:
```diff
@@ -39,25 +42,47 @@ class PytorchInferenceRunner(InferenceRunner):
+      self,
+      batch: List[Union[torch.Tensor, Dict[str, torch.Tensor]]],
+      model: torch.nn.Module,
+      prediction_params: Optional[Dict[str, Any]] = None,
```
Unfortunately, I get this lint warning `Dangerous default value {} as argument (dangerous-default-value)`. 

Some explanations on why setting to `{}` is not good practice:
https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python
https://stackoverflow.com/questions/51710037/how-should-i-use-the-optional-type-hint/51710151#51710151

yeandy replied:
Not relevant anymore due to refactoring into `kwargs`

yeandy commented on sdks/python/apache_beam/ml/inference/pytorch.py:
```diff
@@ -39,25 +42,47 @@ class PytorchInferenceRunner(InferenceRunner):
+          result_dict[k].append(v)
+      for k in result_dict:
+        batched_values = torch.stack(result_dict[k])
+        if batched_values.device != self._device:
```
The original intention was to verify that the examples, which may or may not be attached to GPU during creation time, match the `device` param during `PytorchModelLoader` creation. i.e. a user may pass in `device='GPU'`, and a torch.Tensor(, device='cuda') example, but if GPU is not detected in the environment, then we have to convert the example back to CPU.

yeandy replied:
Added a clarification comment.

yeandy replied:
I can also add a test to verify if this works cleanly.

yeandy commented on sdks/python/apache_beam/ml/inference/pytorch.py:
```diff
@@ -39,25 +42,47 @@ class PytorchInferenceRunner(InferenceRunner):
+        if batched_values.device != self._device:
+          batched_values = batched_values.to(self._device)
+        result_dict[k] = batched_values
+      predictions = model(**result_dict, **prediction_params)
```
Yeah, the naming is not the best. Clarified the variable names.

yeandy commented on sdks/python/apache_beam/ml/inference/pytorch_test.py:
```diff
@@ -59,6 +63,20 @@ def forward(self, x):
+    super().__init__()
+    self.linear = torch.nn.Linear(input_dim, output_dim)
+
+  def forward(self, k1, prediction_param_array, prediction_param_bool):
```
Yes

yeandy commented on sdks/python/apache_beam/ml/inference/pytorch.py:
```diff
@@ -39,25 +42,47 @@ class PytorchInferenceRunner(InferenceRunner):
-    predictions = model(torch_batch)
+    if isinstance(batch[0], dict):
+      result_dict = defaultdict(list)
+      for el in batch:
```
SG, I've cleaned it up now.

yeandy commented on sdks/python/apache_beam/ml/inference/pytorch.py:
```diff
@@ -39,25 +42,47 @@ class PytorchInferenceRunner(InferenceRunner):
+          batched_values = batched_values.to(self._device)
+        result_dict[k] = batched_values
+      predictions = model(**result_dict, **prediction_params)
+    else:
```
Done

yeandy commented on sdks/python/apache_beam/ml/inference/sklearn_inference.py:
```diff
@@ -42,8 +44,12 @@ class ModelFileType(enum.Enum):
+      self,
+      batch: List[numpy.ndarray],
+      model: Any,
+      prediction_params: Optional[Dict[str, Any]] = None
```
SG. Reverted per your other feedback.

yeandy commented on sdks/python/apache_beam/ml/inference/pytorch_test.py:
```diff
@@ -122,6 +137,94 @@ def test_inference_runner_multiple_tensor_features(self):
+    inference_runner = PytorchInferenceRunner(torch.device('cpu'))
+    predictions = inference_runner.run_inference(examples, model)
+    for actual, expected in zip(predictions, expected_predictions):
+      self.assertTrue(_compare_prediction_result(actual, expected))
```
I've changed this to a simple `self.assertEqual(actual, expected)`

yeandy commented on sdks/python/apache_beam/ml/inference/pytorch_test.py:
```diff
@@ -43,10 +43,14 @@
-  return (
-      torch.equal(a.inference, b.inference) and
-      torch.equal(a.example, b.example))
+def _compare_prediction_result(x, y):
```
As opposed to the other pattern
```
for actual, expected in zip(predictions, expected_predictions):
  self.assertTrue(_compare_prediction_result(actual, expected))
```
which will won't return anything meaningful besides saying that `False` is not `True` (which I've already fixed), the call
```
assert_that(
    predictions,
    equal_to(expected_predictions, equals_fn=_compare_prediction_result))
```
will output the details of how `a` and `b` are not equal. So I will keep this function name the same.

yeandy commented:
Updated code per feedback. PTAL @ryanthompson591 @TheNeuralBit

ryanthompson591 commented on sdks/python/apache_beam/ml/inference/pytorch_test.py:
```diff
@@ -59,6 +63,23 @@ def forward(self, x):
+  # k1 is the batched input, and prediction_param_array, prediction_param_bool
+  # are non-batchable inputs (typically model-related info) used to configure
+  # the model before its predict call is invoked
+  def forward(self, k1, k2, prediction_param_array, prediction_param_bool):
```
why not just name the variables what they are instead of k1 and k2 and/or move the hash tag comments as doc for the method.

ryanthompson591 commented on sdks/python/apache_beam/ml/inference/pytorch_test.py:
```diff
@@ -117,11 +135,104 @@ def test_inference_runner_multiple_tensor_features(self):
       self.assertTrue(_compare_prediction_result(actual, expected))
 
+  def test_inference_runner_prediction_params(self):
+    examples = [
```
there's a lot of repeated code here. Consider using a helper to build these examples.

yeandy replied:
Refactored.

ryanthompson591 commented on sdks/python/apache_beam/ml/inference/pytorch_test.py:
```diff
@@ -117,11 +135,104 @@ def test_inference_runner_multiple_tensor_features(self):
+    for actual, expected in zip(predictions, expected_predictions):
+      self.assertEqual(actual, expected)
+
+  def test_inference_runner_kwargs(self):
```
I suggest a short explanation at the head of these tests, since they're more complex than simple unit test.

ryanthompson591 reviewed (commented):
Looking better.

yeandy commented on sdks/python/apache_beam/ml/inference/sklearn_inference.py:
```diff
@@ -42,8 +42,8 @@ class ModelFileType(enum.Enum):
-  def run_inference(self, batch: List[numpy.ndarray],
-                    model: Any) -> Iterable[PredictionResult]:
+  def run_inference(self, batch: List[numpy.ndarray], model: Any,
+                    **kwargs) -> Iterable[PredictionResult]:
```
Not needed. Fixed.

yeandy replied:
Actually, I get this error
```
error: Signature of "run_inference" incompatible with supertype "InferenceRunner"
```
I think that happens when I take out `**kwargs`. Let me add it back and see if the error persists.

ryanthompson591 commented on sdks/python/apache_beam/ml/inference/base_test.py:
```diff
@@ -39,7 +39,7 @@ class FakeInferenceRunner(base.InferenceRunner):
     self._mock_clock = clock
 
-  def run_inference(self, batch: Any, model: Any) -> Iterable[Any]:
+  def run_inference(self, batch: Any, model: Any, **kwargs) -> Iterable[Any]:
```
I suggest a small unit test just to make sure these parameters get passed through to the inference runner.

robertwb commented:
BTW, this is a case where schema'd rows would be very natural.

yeandy commented:
R: @tvalentyn

yeandy commented:
retest this please

yeandy commented:
PTAL @TheNeuralBit

tvalentyn commented:
I took a closer look today. It seems that this PR combines 3 different efforts:
1) Support for batched keyed examples. The key-batched values are passed to pytorch model via `kwargs` in model() call.
2) Support for non-batchable `kwargs` params.
3) Add an interpreted parameter `prediction_params`.

There is collusion of `kwargs` meaning in 1 and 2.  We could explain in the PR description that these are independent efforts (if my understanding is correct).

While I can squash all the changes into one commit before merge, for those looking up the commit history later, it may be easier to follow if efforts 1-3 were separate commits in the repo. Do you want to squash commits accordingly?

yeandy commented:
Makes sense. We can do I've squashed the commits related to 1 into a single commit. Items 2 and 3 are closely interrelated, so I squashed them into a second commit in a new PR here https://github.com/apache/beam/pull/21733. This is done in a new PR because to 1) preserve history in this PR, 2) make it a cleaner rebase

tvalentyn commented:
I think coverage is failing because during the coverage run we don't have pytorch installed and the test is skipped.

Merged the new PR.

tvalentyn commented:
Thanks a lot everyone. thanks @yeandy for cleaning up the commit history.

yeandy commented:
Thanks @tvalentyn for the feedback and merge.
